rust/kernel/virtio.rs

```diff
@@ -0,0 +1,187 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Virtio.
+//!
+//! C header: [`include/linux/virtio.h`](../../../../include/linux/virtio.h)
+
+use crate::{
+    bindings, device, driver, error::from_kernel_result, str::CStr, to_result,
+    types::PointerWrapper, Result, ThisModule,
+};
+
+/// A registration of a platform driver.
+pub type Registration<T> = driver::Registration<Adapter<T>>;
+
+/// Id of an Virtio device.
+#[derive(Clone, Copy)]
+pub struct DeviceId {
+    /// Device id.
+    pub device: u32,
+
+    /// Mask that identifies which bits are valid in the device id.
+    pub vendor: u32,
+}
+
+// SAFETY: `ZERO` is all zeroed-out and `to_rawid` stores in `virtio_id::data`.
+unsafe impl const driver::RawDeviceId for DeviceId {
+    type RawType = bindings::virtio_device_id;
+    const ZERO: Self::RawType = bindings::virtio_device_id {
+        device: 0,
+        vendor: 0,
+    };
+
+    fn to_rawid(&self, _offset: isize) -> Self::RawType {
+        let mut id = Self::ZERO;
+        id.device = self.device;
+        id.vendor = self.vendor;
+        id
+    }
+}
+
+/// An adapter for the registration of platform drivers.
+pub struct Adapter<T: Driver>(T);
+
+impl<T: Driver> driver::DriverOps for Adapter<T> {
+    type RegType = bindings::virtio_driver;
+
+    unsafe fn register(
+        reg: *mut bindings::virtio_driver,
+        name: &'static CStr,
+        module: &'static ThisModule,
+    ) -> Result {
+        // SAFETY: By the safety requirements of this function (defined in the trait definition),
+        // `reg` is non-null and valid.
+        let vdrv = unsafe { &mut *reg };
+
+        vdrv.driver.name = name.as_char_ptr();
+        vdrv.driver.owner = module.0;
+        vdrv.probe = Some(Self::probe_callback);
+        vdrv.remove = Some(Self::remove_callback);
+        if let Some(t) = T::ID_TABLE {
+            vdrv.id_table = t.as_ref();
+        }
+        // SAFETY:
+        //   - `vdrv` lives at least until the call to `register_virtio_driver()` returns.
+        //   - `name` pointer has static lifetime.
+        //   - `module.0` lives at least as long as the module.
+        //   - `probe()` and `remove()` are static functions.
+        //   - `virtio_device_id` is either a raw pointer with static lifetime,
+        //      as guaranteed by the [`driver::IdTable`] type, or null.
+        to_result(unsafe { bindings::register_virtio_driver(reg) })
+    }
+
+    unsafe fn unregister(reg: *mut bindings::virtio_driver) {
+        // SAFETY: By the safety requirements of this function (defined in the trait definition),
+        // `reg` was passed (and updated) by a previous successful call to
+        // `platform_driver_register`.
```

Comment:
Do you mean `register_virtio_driver`?

Reply:
Yes. I will change it in the new commit.